### ISSUE TITLE
[Docs] Add a post-training user guide for the RL-Kernel ablation matrix

### ISSUE THREAD

**Issue by inaniloquentee** [RL-Align/vime#45]
## User story

I am using vime as a post-training user and want to answer a practical question such as:

> Which train/rollout operator or numeric boundary is responsible for the observed logprob mismatch, and what is the performance cost of replacing it with the RL-Kernel path?

At the moment, the repository has the architectural roadmap and individual integration documentation, but it does not have one end-to-end, copy/paste-oriented guide for running and interpreting the post-training ablation matrix. A new user has to infer the matrix semantics from scripts, environment variables, route logs, and generated artifacts. This makes it easy to run an invalid comparison, miss a silent fallback, or report a GPU-time number as an end-to-end speedup.

This issue is intended as an operational documentation follow-up to the RL-Kernel integration roadmap (#6), the strict/audit/auto/off integration work (#43), and the standard alignment adapter (#34). It is not asking for a new orchestration layer.

## What the guide should enable

A user should be able to:

1. Prepare a known-good vime + RL-Kernel post-training environment.
2. Run a near-aligned baseline before changing any operator.
3. Run an operator-isolated train/rollout ablation matrix.
4. Verify the requested implementation was actually used on the training and rollout sides.
5. Reject runs with an unexpected fallback, incomplete artifacts, or a failed job.
6. Compare active-token logprob drift and wall-clock performance correctly.
7. Determine whether the observed difference is caused by an operator, a collective/reduction, metadata, a representation change, or framework scheduling.

The guide should explicitly distinguish a bootstrap/smoke run from a measured post-training run. Smoke runs may validate that a configuration starts, but must not be used as the performance result.

## Matrix semantics that need to be documented

Please define the notation in one stable, user-facing place:

- `P` = production/native implementation.
- `R` = RL-Kernel implementation.
- `P/R` = training uses production/native, rollout uses RL-Kernel.
- `R/P` = training uses RL-Kernel, rollout uses production/native.
- `P/P` = both sides use production/native.
- `R/R` = both sides use RL-Kernel and should be included explicitly as the aligned baseline.

The guide should explain that the first character describes the training side and the second character describes the rollout side. This is important because users otherwise read `P/R` as a single implementation choice rather than a two-sided comparison.

For a module-isolated matrix, only one operator should vary while all other operators remain fixed. For example:

| Matrix | Varied module | Cases | Other modules |
| --- | --- | --- | --- |
| Attention | attention | `R/R`, `P/R`, `R/P`, `P/P` | FFN=`R/R`, logp=`R/R` |
| FFN | FFN | `R/R`, `P/R`, `R/P`, `P/P` | Attention=`R/R`, logp=`R/R` |
| Selected logprob | logp | `R/R`, `P/R`, `R/P`, `P/P` | Attention=`R/R`, FFN=`R/R` |

The document should say when a reduced matrix is acceptable and when a full pairwise matrix is required. It should also explain that `P/P` is not automatically a ground truth for every numerical question: the matrix is for attribution, while the strict consistency contract defines the correctness criterion.

## Required prerequisites and controlled variables

The guide should provide a hardware- and path-agnostic setup, with placeholders instead of one developer's absolute paths. At minimum it should cover:

- vime, RL-Kernel, Megatron, PyTorch, CUDA, vLLM, and vllm-router versions;
- model checkpoint identity and weight version;
- tokenizer and chat-template identity;
- prompt/data file and field names;
- training/rollout topology, TP/CP/PP/EP, GPU allocation, and colocated versus disaggregated mode;
- dtype, quantization, KV-cache policy, attention backend, and logprob backend;
- rollout temperature, top-p/top-k, response length, batch sizes, dynamic batching, and packing policy;
- seeds and warmup policy;
- weight-sync mode and the point at which rollout samples are compared with training replay;
- process-start and build-time variables versus request-time variables.

The guide should require that baseline and candidate runs use the same semantic inputs and model state. It should warn that changing sampling filters, checkpoint timing, token shift/masks, padding/packing, TP/CP layout, or weight-sync timing at the same time as changing an operator invalidates operator attribution.

## Suggested user-facing workflow

A documented workflow could look like this (with paths and supported entry points adapted to the public vime interface):

```bash
export VIME_ROOT=/path/to/vime
export RL_KERNEL_ROOT=/path/to/RL-Kernel
export PYTHON_BIN=/path/to/the/validated/python
export MATRIX_ROOT="$VIME_ROOT/outputs/rlkernel/ablation-matrix-$STAMP"

# 1. Run one near-aligned R/R baseline and keep its config/provenance.
# 2. Run the attention, FFN, and selected-logprob matrix entries.
# 3. Analyze all entries and fail if a route or artifact is incomplete.
bash "$VIME_ROOT/scripts/run-ablation-matrix.sh"
"$PYTHON_BIN" "$VIME_ROOT/scripts/analyze-ablation-matrix.py" \
  --matrix-root "$MATRIX_ROOT" \
  --output "$MATRIX_ROOT/matrix-report.json"
```

If the current scripts are still experimental, the documentation should say so and either make their path/configuration public and parameterized, or provide a supported replacement. A user should not need to edit a shell script to change the repository root, Python environment, checkpoint, data path, run stamp, or output directory.

The guide should document the mode controls and their relationship to the matrix, including the currently used strict/audit/auto/off behavior and any preferred CLI flags. It should make clear which knobs are supported public API and which are developer-only experiment overrides such as per-module case variables.

## Artifacts and validity checks

Please document the expected output layout and what each file proves. A useful minimum is:

```text
<matrix-root>/
  <module>-<case>/
    matrix-case.env
    exit-code
    run.log
    readbacks/*.json
    train-data/                 # optional, when replay/debug data is enabled
    bitwise-validation.json     # when strict validation is enabled
  matrix-report.json
```

For every matrix entry, the analyzer should report at least:

- requested matrix case and actual applied case;
- training-side and rollout-side implementation/backend IDs;
- contract ID and build/runtime fingerprint when available;
- fallback count and structured fallback reason;
- job exit status and traceback/CUDA/NCCL/HTTP/OOM errors;
- active-token count and mask coverage;
- `train/train_rollout_logprob_abs_diff` or the current canonical dlogp metrics;
- mismatch count, maximum absolute drift, and percentile statistics when available;
- rollout time, training time, weight-sync time, total step time, and tokens/sec;
- artifact completeness and the final pass/fail decision.

A run should not be considered a valid ablation merely because the process exits successfully. An `auto` run that silently falls back is not evidence for the requested optimized implementation; it must either be labeled as a fallback case or fail the strict matrix validator.

## How results should be interpreted

The documentation should separate four questions:

1. **Did the intended route execute?** Check requested versus actual backend, readbacks, and fallback records.
2. **Did the semantic comparison remain valid?** Check checkpoint/weight version, tokens, masks, sampling metadata, padding/layout, and topology.
3. **Did the numeric contract pass?** Report active-token dlogp metrics using the RL-Kernel contract as the source of truth; do not duplicate tolerance constants in vime docs.
4. **Did the run get faster end to end?** Compare wall-clock phase timers over repeated measured runs. Do not treat the sum of GPU kernel time across processes as step time, and do not rank persistent synchronization kernels from a full-job trace without phase attribution.

The guide should include a result table template with baseline, candidate, implementation route, drift metrics, rollout time, training time, weight-sync time, total step time, memory, and notes. It should recommend reporting a mean/median and run-to-run variation rather than a single noisy one-round number when making a performance claim.

It should also explain the attribution ladder:

1. semantic input and metadata;
2. single-device arithmetic;
3. distributed reduction/topology;
4. representation/dtype/quantization;
5. framework scheduling and overlap;
6. pairwise operator interactions;
7. production end-to-end run.

This prevents a user from interpreting a train-rollout mismatch caused by top-p processed-logprob semantics or weight-sync timing as a CUDA operator failure.

## Troubleshooting section requested

Please include short recipes for the most common failures:

- expected backend is not present in readbacks;
- strict mode fails because a provider or extension is unavailable;
- auto mode silently falls back;
- `max_abs_dlogp` is non-zero but token/mask metadata is not identical;
- R/R is stable but P/R or R/P drifts;
- CUDA Graph or eager mode changes the observed route;
- a run completes but has missing `readbacks` or debug tensors;
- GPU time is high while wall-clock time is unchanged because the kernel is overlapped;
- the matrix changes after a vLLM restart or weight update;
- performance results are polluted by profiler, checkpoint writing, or debug readback overhead.

## Proposed acceptance criteria

- Add linked English and Chinese documentation under an appropriate advanced/developer section.
- Define `P/R`, `R/P`, `P/P`, and `R/R` unambiguously, including which side each character refers to.
- Provide a copy/paste workflow that works from a clean checkout after replacing documented placeholders.
- Include an explicit R/R baseline and a single-variable operator matrix.
- Document all required fixed variables and the distinction between request-time, process-start, and build-time controls.
- Document actual-route/fallback validation and the artifact directory schema.
- Provide a result table and explain wall-clock versus cross-process GPU-time accounting.
- State what strict bitwise/tolerance claims are made at operator, train-rollout, and multi-round training levels.
- Link the guide from the existing RL-Kernel integration and consistency documentation.
- Include at least one checked-in example report or a small CI/manual validation recipe so the documentation stays executable.

The goal is to make the matrix usable by a post-training practitioner who did not write the integration, while keeping vime responsible for orchestration and RL-Kernel responsible for its backend contracts.
## Additional requirement: MI330X / AMD ROCm operating guide

The same post-training ablation workflow should also be documented for AMD MI330X systems. A CUDA/H100-only procedure is not sufficient for a post-training user because the runtime, communication library, device architecture, and supported graph/IPC behavior can differ substantially on ROCm.

The AMD section should be a first-class variant of the same workflow, not a separate correctness definition. The matrix semantics must remain identical:

- `P` = the production/native implementation available in the validated ROCm stack;
- `R` = the RL-Kernel implementation, only when the backend explicitly reports MI330X/ROCm support;
- `P/R`, `R/P`, `P/P`, and `R/R` retain the same training-side/rollout-side meaning as in the main guide.

### MI330X prerequisites and environment fingerprint

Please provide an MI330X/ROCm checklist with placeholders rather than assuming the CUDA paths or package versions used on H100:

- MI330X model, GPU count, node count, interconnect/topology, and `gfx` target reported by `rocminfo`;
- ROCm, HIP, PyTorch ROCm, vLLM, vllm-router, Vime, Megatron, RCCL, and compiler versions;
- `torch.version.hip`, `torch.cuda.is_available()`, device count, and the resolved vime/RL-Kernel source paths;
- `rocminfo` and `rocm-smi` output captured as run provenance;
- the exact `PYTORCH_ROCM_ARCH`/HIP architecture setting when a native extension is built;
- whether the selected attention, GEMM, logprob, normalization, and communication backends actually support MI330X;
- whether the run is colocated or disaggregated and which weight-sync transport is supported on that topology;
- ROCm/RCCL environment variables and their scope, including which settings are process-start-only;
- memory capacity, allocator configuration, and any MI330X-specific vLLM limits.

The documentation should include a preflight command that verifies the runtime and reports the detected architecture, for example:

```bash
export VIME_ROOT=/path/to/vime
export RL_KERNEL_ROOT=/path/to/RL-Kernel
export PYTHON_BIN=/path/to/validated/rocm/python

"$PYTHON_BIN" -c "import torch, vllm, vime; print('torch', torch.__version__, 'hip', torch.version.hip, 'cuda_api', torch.cuda.is_available(), 'gpus', torch.cuda.device_count()); print('vime', vime.__file__); print('vllm', vllm.__version__)"
rocminfo
rocm-smi
```

The guide should explicitly say that `torch.cuda` is still the PyTorch device API on ROCm, while CUDA-specific assumptions in scripts, extension names, graph support, and IPC paths must not be inferred from that API name.

### MI330X matrix and fallback rules

The AMD workflow should use the same one-variable-at-a-time matrix:

| Matrix | Varied module | Cases | Other modules |
| --- | --- | --- | --- |
| Attention | attention | `R/R`, `P/R`, `R/P`, `P/P` | FFN=`R/R`, logp=`R/R` |
| FFN | FFN | `R/R`, `P/R`, `R/P`, `P/P` | Attention=`R/R`, logp=`R/R` |
| Selected logprob | logp | `R/R`, `P/R`, `R/P`, `P/P` | Attention=`R/R`, FFN=`R/R` |

For each cell, the documentation must show the requested case, the actual backend, and the expected behavior when MI330X support is incomplete:

- `strict`: fail closed with a clear unsupported-ROCm/MI330X reason;
- `audit`: run only the available path and record that the requested route was not executed;
- `auto`: allow fallback, but label the cell as a fallback and exclude it from an RL-Kernel performance claim;
- `off`: provide the native ROCm baseline.

An unavailable RL-Kernel backend must not be silently presented as `R`, and a production ROCm fallback must not be counted as an optimized result. The report should distinguish unsupported backend, build failure, runtime fallback, and semantic mismatch.

### AMD-specific correctness and performance guidance

The AMD section should call out the following points for users:

1. Keep checkpoint, tokenizer, token IDs, masks, sampling parameters, response length, batch shape, seeds, TP/CP/PP/EP layout, weight version, and update timing fixed across cells.
2. Validate semantic identity before attributing a difference to HIP kernels or RCCL.
3. Record the exact ROCm architecture and build fingerprint because a kernel compiled for one `gfx` target should not be assumed to represent another AMD target.
4. Do not assume CUDA Graph, HIP Graph, CUDA IPC, or peer-memory behavior is equivalent. Each must be reported as supported, unavailable, or not tested for the exact ROCm version and topology.
5. Treat RCCL communication selection and environment variables as an explicit matrix axis only when intentionally testing them. A NCCL algorithm setting copied from an NVIDIA run must not be described as an effective AMD setting without runtime evidence.
6. Measure MI330X performance against an MI330X baseline. Do not compare raw H100 and MI330X wall-clock numbers as an operator speedup without normalizing model, workload, software versions, and topology.
7. Keep profiler and debug readback overhead out of the measured run, and report wall-clock phase timers separately from cross-process GPU-kernel totals.

The result report should include at least:

- `torch.version.hip`, ROCm/HIP/RCCL versions, `gfx` target, and GPU topology;
- requested/actual implementation and backend IDs for training and rollout;
- contract ID, build fingerprint, fallback reason, and capability decision;
- active-token logprob drift, mismatch count, maximum/percentile error, and mask coverage;
- rollout, training, weight-sync, and total step wall-clock times;
- tokens/sec, peak memory, OOM/illegal-access/NCCL-or-RCCL/HTTP errors;
- whether the run used eager execution, graph capture, peer memory, or a fallback communication path.

### MI330X acceptance criteria

The documentation work should be considered complete when a user with a clean MI330X/ROCm checkout can:

- run the documented preflight and see the complete software/hardware fingerprint;
- run the native ROCm `P/P` baseline and an explicit `R/R` probe when RL-Kernel support is available;
- run the attention, FFN, and selected-logprob matrix cells without editing developer-only absolute paths;
- see a structured, fail-closed result when a backend is unsupported instead of an ambiguous import error or silent fallback;
- reproduce the same artifact layout and report schema as the CUDA workflow;
- distinguish correctness/contract results from performance results;
- understand which parts of the CUDA workflow are unavailable or require an AMD-specific implementation.

If MI330X support is not yet implemented for a particular RL-Kernel operator, the guide should still document the matrix and the expected unsupported result. It should not claim `R/R` coverage until the backend, numeric contract, and actual route have been validated on MI330X.
